src/main/java/org/elasticsearch/rest/BytesRestResponse.java

```diff
@@ -120,9 +120,21 @@ public RestStatus status() {
     }
 
     private static XContentBuilder convert(RestChannel channel, RestStatus status, Throwable t) throws IOException {
-        XContentBuilder builder = channel.newBuilder().startObject()
-                .field("error", detailedMessage(t))
-                .field("status", status.getStatus());
+        XContentBuilder builder = channel.newBuilder().startObject();
+        if (t == null) {
+            builder.field("error", "Unknown");
+        } else if (channel.detailedErrorsEnabled()) {
+            builder.field("error", detailedMessage(t));
+            buildErrorTrace(channel, t, builder);
+        } else {
+            builder.field("error", simpleMessage(t));
+        }
+        builder.field("status", status.getStatus());
+        builder.endObject();
+        return builder;
+    }
+
+    private static void buildErrorTrace(RestChannel channel, Throwable t, XContentBuilder builder) throws IOException {
```

Comment:
maybe move the conditional logic back out to the caller? the null check isn't needed, and the function name implies it always adds this trace, when in fact it depends on the request.